### PR TITLE
[418] Add the support for list container

### PR DESCRIPTION
**Pull request by gcoutable**
- Creates a node list when the Sirius container mapping style is "LIST"
- Creates a node list item when a Sirius node mapping style represents a
simple figure (ellipse, lozenge, square)
- Improves ELK configuration to support these node types
- Updates ELK dependency to 0.7.1 because a bug prevents
NodeSizeConstraints option to be correctly applied
- Provides an OS aware minimum size for empty labels.

Bug: eclipse-sirius/sirius-components#418
Signed-off-by: Guillaume Coutable <guillaume.coutable@obeo.fr>

### Type of this PR 

- [ ] Bug fix
- [x] New feature or improvement
- [ ] Documentation
- [ ] Cleanup
- [ ] Test
- [ ] Build/Releng

### Issue(s)

https://github.com/eclipse-sirius/sirius-components/issues/418

### What does this PR do?

This PR provides the first support of NodeList and NodeListItem for sirius-component with the ELK configuration.

### Screenshot/screencast of this PR

...
 

### Potential side effects

Does not work well with the incremental layout yet.

### How to test this PR?

- [ ] Frontend Unit tests
- [ ] Backend Unit tests
- [ ] Cypress : please specify
- [x] Manual Test : 
  - From the Model explorer :
    - Creates an empty model
    - Creates a Domain
    - Creates an Entity
    - Creates many attributes (event with empty labels)
    - Creates a _Domain_ representation

### Checklist

- [x] I have read CONTRIBUTING carefully.
- [x] I have signed the [Contributor License Agreement](http://www.eclipse.org/legal/CLA.php).
- [x] All my commits are signed-off (`-s`) with my mail address of my Eclipse Account.
- [ ] I have covered my changes by unit tests or integration tests or manual tests.
- [ ] All tests pass.
- [ ] I have updated the documentation accordingly
- [ ] New React components are availables in storybook

### CODE REVIEW

File: backend/sirius-web-compatibility/src/main/java/org/eclipse/sirius/web/compat/diagrams/NodeMappingStyleProvider.java [hunk 1]
```diff
@@ -51,7 +58,9 @@ public INodeStyle apply(VariableManager variableManager) {
     private INodeStyle getNodeStyle(VariableManager variableManager, NodeStyleDescription nodeStyleDescription) {
         INodeStyle style = null;
 
-        if (nodeStyleDescription instanceof SquareDescription) {
+        if (this.shouldBeConsideredAsListItemNodeStyle(variableManager, nodeStyleDescription)) {
+            style = this.createContainerListItemNodeStyle(variableManager, nodeStyleDescription);
```
Comment (sbegaudeau) [inline backend/sirius-web-compatibility/src/main/java/org/eclipse/sirius/web/compat/diagrams/NodeMappingStyleProvider.java:62]: No need to talk about container here

File: backend/sirius-web-compatibility/src/main/java/org/eclipse/sirius/web/compat/diagrams/NodeMappingStyleProvider.java [hunk 2]
```diff
@@ -75,6 +84,34 @@ private INodeStyle getNodeStyle(VariableManager variableManager, NodeStyleDescri
...
+        if ((nodeStyleDescription instanceof SquareDescription || nodeStyleDescription instanceof EllipseNodeDescription || nodeStyleDescription instanceof LozengeNodeDescription)
+                && this.nodeMapping.eContainer() instanceof ContainerMapping) {
+            ContainerMapping parentMapping = (ContainerMapping) this.nodeMapping.eContainer();
+            ContainerStyleDescription containerStyleDescription = new ContainerStyleDescriptionProvider(this.interpreter, parentMapping).getContainerStyleDescription(variableManager);
+            return containerStyleDescription instanceof FlatContainerStyleDescription && ContainerLayout.LIST.equals(parentMapping.getChildrenPresentation());
+        }
+        return false;
+    }
+
+    private INodeStyle createContainerListItemNodeStyle(VariableManager variableManager, NodeStyleDescription nodeStyleDescription) {
```
Comment (sbegaudeau) [inline backend/sirius-web-compatibility/src/main/java/org/eclipse/sirius/web/compat/diagrams/NodeMappingStyleProvider.java:107]: Container can be removed

File: backend/sirius-web-diagrams-layout/src/main/java/org/eclipse/sirius/web/diagrams/layout/TextBoundsService.java [hunk 3]
```diff
@@ -90,4 +95,15 @@ public TextBounds getBounds(Label label) {
...
+    public TextBounds getTextMinimumBounds() {
+        // @formatter:off
+        LabelStyle labelStyle = LabelStyle.newLabelStyle()
+                .fontSize(16)
+                .color("#000000") //$NON-NLS-1$
+                .iconURL("") //$NON-NLS-1$
+                .build();
+        // @formatter:on
+        return this.textBoundsProvider.computeBounds(labelStyle, MINIMUM_TEXT_LABEL_CHARACTERS);
+    }
```
Comment (sbegaudeau) [inline backend/sirius-web-diagrams-layout/src/main/java/org/eclipse/sirius/web/diagrams/layout/TextBoundsService.java]: You can use a constant for that instead of computing a hardcoded value like that.

File: backend/sirius-web-diagrams-layout/src/main/java/org/eclipse/sirius/web/diagrams/layout/TextBoundsService.java [hunk 4]
```diff
@@ -38,6 +38,11 @@
 @Service
 public class TextBoundsService {
 
+    /**
+     * The text label minimum characters needed to consider an empty node easily clickable.
+     */
+    private static final String MINIMUM_TEXT_LABEL_CHARACTERS = "    "; //$NON-NLS-1$
+
```
Comment (sbegaudeau) [inline backend/sirius-web-diagrams-layout/src/main/java/org/eclipse/sirius/web/diagrams/layout/TextBoundsService.java:45]: Can be removed

File: backend/sirius-web-diagrams-layout/src/main/java/org/eclipse/sirius/web/diagrams/layout/LayoutConfiguratorRegistry.java [hunk 5]
```diff
@@ -62,7 +67,10 @@
 
     private final List<IDiagramLayoutConfiguratorProvider> customLayoutProviders;
 
-    public LayoutConfiguratorRegistry(List<IDiagramLayoutConfiguratorProvider> customLayoutProviders) {
+    private TextBoundsService textBoundsService;
+
+    public LayoutConfiguratorRegistry(List<IDiagramLayoutConfiguratorProvider> customLayoutProviders, TextBoundsService textBoundsService) {
+        this.textBoundsService = textBoundsService;
```
Comment (sbegaudeau) [inline backend/sirius-web-diagrams-layout/src/main/java/org/eclipse/sirius/web/diagrams/layout/LayoutConfiguratorRegistry.java:64]: Can be removed

File: backend/sirius-web-diagrams-layout/src/main/java/org/eclipse/sirius/web/diagrams/layout/LayoutConfiguratorRegistry.java [hunk 6]
```diff
@@ -83,6 +91,23 @@ public LayoutConfigurator getDefaultLayoutConfigurator() {
...
+                .setProperty(CoreOptions.NODE_SIZE_CONSTRAINTS, SizeConstraint.free())
+                .setProperty(CoreOptions.NODE_SIZE_MINIMUM, new KVector(MIN_WIDTH_CONSTRAINT, MIN_HEIGHT_CONSTRAINT))
+                .setProperty(CoreOptions.PADDING, new ElkPadding(6d, 12d, 12d, 0))
+                .setProperty(CoreOptions.NODE_LABELS_PADDING, new ElkPadding())
+                .setProperty(CoreOptions.SPACING_NODE_NODE, 6d)
+                .setProperty(LayeredOptions.CONSIDER_MODEL_ORDER, OrderingStrategy.NODES_AND_EDGES)
+                .setProperty(CoreOptions.NODE_LABELS_PLACEMENT, NodeLabelPlacement.insideTopCenter());
+
+
+        Size minimumLabelSize = this.textBoundsService.getTextMinimumBounds().getSize();
```
Comment (sbegaudeau) [inline backend/sirius-web-diagrams-layout/src/main/java/org/eclipse/sirius/web/diagrams/layout/LayoutConfiguratorRegistry.java:95]: Can be removed

File: frontend/package.json [hunk 7]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@eclipse-sirius/sirius-components",
-  "version": "0.2.18",
+  "version": "0.2.19",
```
Comment (sbegaudeau) [inline frontend/package.json:3]: Do not upgrade that this way (have a look at the versioning ADR for the proper commands for that and do it in another commit too since it changes a lot of files)

Comment (gcoutable) [reply frontend/package.json]: Ack

File: backend/sirius-web-diagrams-layout/src/main/java/org/eclipse/sirius/web/diagrams/layout/LayoutConfiguratorRegistry.java (hunk 6 above)
Comment (gcoutable) [reply backend/sirius-web-diagrams-layout/src/main/java/org/eclipse/sirius/web/diagrams/layout/LayoutConfiguratorRegistry.java]: Done

File: backend/sirius-web-diagrams-layout/src/main/java/org/eclipse/sirius/web/diagrams/layout/LayoutConfiguratorRegistry.java (hunk 5 above)
Comment (gcoutable) [reply backend/sirius-web-diagrams-layout/src/main/java/org/eclipse/sirius/web/diagrams/layout/LayoutConfiguratorRegistry.java]: Done

File: backend/sirius-web-diagrams-layout/src/main/java/org/eclipse/sirius/web/diagrams/layout/TextBoundsService.java (hunk 4 above)
Comment (gcoutable) [reply backend/sirius-web-diagrams-layout/src/main/java/org/eclipse/sirius/web/diagrams/layout/TextBoundsService.java]: Done

File: backend/sirius-web-diagrams-layout/src/main/java/org/eclipse/sirius/web/diagrams/layout/TextBoundsService.java (hunk 3 above)
Comment (gcoutable) [reply backend/sirius-web-diagrams-layout/src/main/java/org/eclipse/sirius/web/diagrams/layout/TextBoundsService.java]: No need for a constant anymore

File: backend/sirius-web-compatibility/src/main/java/org/eclipse/sirius/web/compat/diagrams/NodeMappingStyleProvider.java (hunk 2 above)
Comment (gcoutable) [reply backend/sirius-web-compatibility/src/main/java/org/eclipse/sirius/web/compat/diagrams/NodeMappingStyleProvider.java]: done

File: backend/sirius-web-compatibility/src/main/java/org/eclipse/sirius/web/compat/diagrams/NodeMappingStyleProvider.java (hunk 1 above)
Comment (gcoutable) [reply backend/sirius-web-compatibility/src/main/java/org/eclipse/sirius/web/compat/diagrams/NodeMappingStyleProvider.java]: Done

File: backend/sirius-web-diagrams-layout/src/main/java/org/eclipse/sirius/web/diagrams/layout/LayoutConfiguratorRegistry.java [hunk 8]
```diff
@@ -62,7 +54,7 @@
 
     private final List<IDiagramLayoutConfiguratorProvider> customLayoutProviders;
 
-    public LayoutConfiguratorRegistry(List<IDiagramLayoutConfiguratorProvider> customLayoutProviders) {
+    public LayoutConfiguratorRegistry(List<IDiagramLayoutConfiguratorProvider> customLayoutProviders, TextBoundsService textBoundsService) {
```
Comment (sbegaudeau) [inline backend/sirius-web-diagrams-layout/src/main/java/org/eclipse/sirius/web/diagrams/layout/LayoutConfiguratorRegistry.java:57]: Can be removed

File: backend/sirius-web-diagrams-layout/src/main/java/org/eclipse/sirius/web/diagrams/layout/ELKDiagramConverter.java [hunk 9]
```diff
@@ -74,16 +79,225 @@ public ELKDiagramConverter(TextBoundsService textBoundsService, ImageSizeProvide
...
+
+        // @formatter:off
+         return Arrays.asList(
+                largestNodeListItemWidth,
+                nodeListLabelTextBounds.getSize().getWidth(),
+                LayoutOptionValues.MIN_WIDTH_CONSTRAINT)
+              .stream()
+              .mapToDouble(Number::doubleValue)
+              .max()
+              .orElse(LayoutOptionValues.MIN_WIDTH_CONSTRAINT);
```
Comment (sbegaudeau) [inline backend/sirius-web-diagrams-layout/src/main/java/org/eclipse/sirius/web/diagrams/layout/ELKDiagramConverter.java:186]: No @formatter:on

**Review by sbegaudeau — CHANGES_REQUESTED** [review]
I don't know why but the resizing does not work anymore

File: backend/sirius-web-compatibility/src/main/java/org/eclipse/sirius/web/compat/diagrams/AbstractNodeMappingStyleProvider.java [hunk 10]
```diff
@@ -79,6 +91,56 @@ private INodeStyle getNodeStyle(VariableManager variableManager, LabelStyleDescr
...
+     * lozenge) and the {@link #nodeMapping} container represents his children in a list, <code>false</code> otherwise.
+     *
+     * @param variableManager
+     *            The variable manager
+     * @param nodeStyleDescription
+     *            the node style description that could be considered as a list item node style
+     * @return <code>true</code> if the <em>nodeStyleDescription</em> should be considered as a list item node style
+     */
+    private boolean shouldBeConsideredAsListItemNodeStyle(VariableManager variableManager, LabelStyleDescription nodeStyleDescription) {
+        if ((nodeStyleDescription instanceof SquareDescription || nodeStyleDescription instanceof EllipseNodeDescription || nodeStyleDescription instanceof LozengeNodeDescription)
```
Comment (pcdavid) [inline backend/sirius-web-compatibility/src/main/java/org/eclipse/sirius/web/compat/diagrams/AbstractNodeMappingStyleProvider.java:105]: Why all these tests, and not just "my parent is a list"?

File: backend/sirius-web-compatibility/src/main/java/org/eclipse/sirius/web/compat/diagrams/AbstractNodeMappingStyleProvider.java [hunk 11]
```diff
@@ -79,6 +91,56 @@ private INodeStyle getNodeStyle(VariableManager variableManager, LabelStyleDescr
...
+        // @formatter:off
+        return ListItemNodeStyle.newListItemNodeStyle()
+                .backgroundColor("transparent") //$NON-NLS-1$
+                .build();
+        // @formatter:on
+    }
+
+    private INodeStyle createListNodeStyle(VariableManager variableManager, FlatContainerStyleDescription flatContainerStyleDescription) {
+        ColorDescriptionConverter backgroundColorProvider = new ColorDescriptionConverter(this.interpreter, variableManager.getVariables());
+        ColorDescriptionConverter borderColorProvider = new ColorDescriptionConverter(this.interpreter, variableManager.getVariables());
```
Comment (pcdavid) [inline backend/sirius-web-compatibility/src/main/java/org/eclipse/sirius/web/compat/diagrams/AbstractNodeMappingStyleProvider.java:124]: No need to create two converters. They have the exact same configuration (and `variableManager.getVariables()` is relatively costly).

File: backend/sirius-web-compatibility/src/main/java/org/eclipse/sirius/web/compat/diagrams/LabelStyleDescriptionConverter.java [hunk 12]
```diff
@@ -37,6 +39,19 @@ public LabelStyleDescriptionConverter(AQLInterpreter interpreter, IObjectService
...
 
+    private String getImagePath(Object object) {
+        if (object instanceof EObject) {
+            EObject eObject = (EObject) object;
+            if (eObject.eClass().getName().equals("Attribute")) { //$NON-NLS-1$
+                Enumerator value = (Enumerator) eObject.eGet(eObject.eClass().getEStructuralFeature("type")); //$NON-NLS-1$
+                if (value.getName().equals("STRING")) { //$NON-NLS-1$
+                    return "/icons/full/obj16/Attribute.gif"; //$NON-NLS-1$
+                }
+            }
```
Comment (pcdavid) [inline backend/sirius-web-compatibility/src/main/java/org/eclipse/sirius/web/compat/diagrams/LabelStyleDescriptionConverter.java:50]: This looks lioke a hard-coded special case for a specific modeler/domain?

File: backend/sirius-web-diagrams-layout/src/main/java/org/eclipse/sirius/web/diagrams/layout/ELKDiagramConverter.java [hunk 13]
```diff
@@ -74,16 +79,225 @@ public ELKDiagramConverter(TextBoundsService textBoundsService, ImageSizeProvide
...
+        for (Node nodeListItem : nodeListItems) {
+            TextBounds nodeListItemTextBounds = this.textBoundsService.getBounds(nodeListItem.getLabel());
+            double nodeListItemWidth = nodeListItemTextBounds.getSize().getWidth() + LayoutOptionValues.NODE_LIST_ELK_NODE_LABELS_PADDING_RIGHT
+                    + LayoutOptionValues.NODE_LIST_ELK_NODE_LABELS_PADDING_LEFT;
+            if (largestNodeListItemWidth < nodeListItemWidth) {
+                largestNodeListItemWidth = nodeListItemWidth;
+            }
+        }
+
+        return largestNodeListItemWidth;
```
Comment (pcdavid) [inline backend/sirius-web-diagrams-layout/src/main/java/org/eclipse/sirius/web/diagrams/layout/ELKDiagramConverter.java:163]: Maybe

```java
nodeListItems.stream().max(Comparator.comparing(item -> this.textBoundsService.getBounds(item.getLabel()))
  + LayoutOptionValues.NODE_LIST_ELK_NODE_LABELS_PADDING_RIGHT + LayoutOptionValues.NODE_LIST_ELK_NODE_LABELS_PADDING_LEFT;
```

**Review by pcdavid — CHANGES_REQUESTED** [review]

File: backend/sirius-web-diagrams-layout/src/main/java/org/eclipse/sirius/web/diagrams/layout/ELKDiagramConverter.java (hunk 9 above)
Comment (gcoutable) [reply backend/sirius-web-diagrams-layout/src/main/java/org/eclipse/sirius/web/diagrams/layout/ELKDiagramConverter.java:186]: Done

**Review by pcdavid — APPROVED** [review]
